rfcs/0007-synchronous-commit.md

```diff
@@ -0,0 +1,162 @@
+
+## Current Design in SlateDB
+
+This section is based on @criccomini 's comment in <https://github.com/slatedb/slatedb/pull/260#issuecomment-2576502212>.
+
+SlateDB currently does not provide an explicit notion of Synchronous Commit. But it does provide a `DurabilityLevel` enum to control the durability guarantees on both read and write operations.
+
+The `DurabilityLevel` enum is defined as follows:
+
+```rust
+enum DurabilityLevel {
+    Memory,
+    Local, // not implemented yet
+    Remote,
+}
+```
+
+And the `WriteOptions` struct contains a `await_durability: DurabilityLevel` option to control the waiting behavior for durability. If `await_durability` is set to `DurabilityLevel::Remote`, the write will wait for the WAL to be written into S3 before returning.
+
+Please note that the commit semantic is a bit different from other systems' Synchronous Commit. No matter what `DurabilityLevel` is set in the write operation, this write is considered visible to the readers with `DurabilityLevel::Memory` immediately after the write is appended to the WAL, not nessarily flushed to storage.
+
+The reason is that SlateDB's WAL is not a place for crash recovery only, but also a place for data reads. The read path is first access the WAL, then MemTable, then L0 SST, then SSTs at deeper levels.
+
+In the notion of Synchronous Commit, the data is considered as committed as soon as the write is persisted to the WAL storage. Users can specify the durability level as `DurabilityLevel::Remote` for the read calls to ensure only the committed/persisted data is read.
+
+SlateDB differs from both PostgreSQL and RocksDB in multiple ways. Unlike PostgreSQL, it's not a distributed system with Primary/Standby nodes. And unlike RocksDB, it stores data in S3 rather than local disk, which means slower write operations and additional costs for API requests. These differences lead to several key considerations:
+
+1. Group commit is essential. By batching multiple writes together, we can reduce both API costs and improve performance compared to multiple small writes. However, even with Group Commit, writes to S3 will still be slower than local disk writes. (While parallel writes to S3 could potentially improve performance, this would increase both API costs and the complexity of handling failures.)
+2. Writes will inherently take longer due to S3 latency. Given this reality, it makes sense to allow readers who can accept eventual consistency to access unpersisted and uncommitted data while waiting for writes to be durably committed.
+3. Writing WAL to S3 has a higher risk of permanent failures due to network instability compared to local disk. This makes it critical to implement robust auto-recovery mechanisms for handling I/O failures.
```

Comment:
Perhaps reference @thrawn01's PRs/RFCs here since he's focusing on corruption and recovery right now.

Reply:
:+1: 

thank you for remiding, it's good to have this reference link here.

besides data corruption, how to deal with network failure / replica failure on sync commit, also contains some details.

postgres's 'remote_apply' and MySQL's semi-synchronous replication semantics might serve as a reference for this, as they also share the characteristic of dealing with network instability during commits. if all replicas go down, the write on primary should fail like nothing happens. however, I'm still unsure about the details regarding how to handle the WAL in this scenario (do they did something like 2pc to revert the WAL?). still need some investigations on this 🤔.